src/ui/components/NetworkMonitor/index.tsx

```diff
@@ -1,51 +1,52 @@
 import SplitBox from "devtools/client/shared/components/splitter/SplitBox";
 import React, { useEffect, useRef, useState } from "react";
-import { connect, ConnectedProps, useDispatch } from "react-redux";
+import { connect, ConnectedProps, useDispatch, useSelector } from "react-redux";
 import { actions } from "ui/actions";
 import {
   getFormattedFrames,
   getResponseBodies,
   getRequestBodies,
   getFocusedEvents,
   getFocusedRequests,
+  getSelectedRequestId,
 } from "ui/reducers/network";
 import { getCurrentTime } from "ui/reducers/timeline";
 import { UIState } from "ui/state";
-import RequestDetails, { RequestDetailsUnavailable } from "./RequestDetails";
+import RequestDetails from "./RequestDetails";
 import RequestTable from "./RequestTable";
 import { CanonicalRequestType, RequestSummary } from "./utils";
 import FilterBar from "./FilterBar";
 import Table from "./Table";
-import { fetchResponseBody, fetchRequestBody } from "ui/actions/network";
+import {
+  fetchResponseBody,
+  fetchRequestBody,
+  hideRequestDetails,
+  showRequestDetails,
+} from "ui/actions/network";
 import { getThreadContext } from "devtools/client/debugger/src/selectors";
 import LoadingProgressBar from "../shared/LoadingProgressBar";
 import { trackEvent } from "ui/utils/telemetry";
 import { timeMixpanelEvent } from "ui/utils/mixpanel";
 import { getLoadedRegions } from "ui/reducers/app";
-import { getPointIsInLoadedRegion } from "ui/utils/timeline";
 
 export const NetworkMonitor = ({
   currentTime,
   cx,
   events,
-  frames,
   loading,
-  loadedRegions,
   requestBodies,
   requests,
   responseBodies,
   seek,
   selectFrame,
 }: PropsFromRedux) => {
-  const [selectedRequest, setSelectedRequest] = useState<RequestSummary>();
+  const selectedRequestId = useSelector(getSelectedRequestId);
```

Comment:
When this was state that was local to this component, it made sense to pass it through to its child `RequestDetails`. Now that this is coming from the global store, I think it would be better to fetch it separately from each component, so that they can be more self-contained (e.g. in the future, you could render a `RequestDetails` somewhere else and you as the renderer would not need to go asking the global store for the request ID to do that). It means duplicating this line across the two components, but I think it's closer to the `many small connected components` idea which is the modern norm (calling on @markerikson to correct me if I'm wrong on that one 😉 )

Reply:
Hmm. We definitely encourage more components reading their own data from the store as a general principle.  In this case, though, I'm not sure we _want_ to do that.

Reading data from the store forces a component to re-render when that data changes. I'm not sure of the component hierarchy in this area of the UI, but it sounds list-y.

Ideally, if I have list item 4 selected, then click on list item 7, _only_ a max of three components render: item 4, item 7, and likely their parent list.

If the list item component reads a "current selected ID" value, now _every_ list item will re-render whenever that changes.  

But, if we read that "current selected ID" value in the list _parent_, the same thing happens because React re-renders recursively by default.

So, _if_ we want to optimize re-rendering, the best approach here would be:

- Parent reads "current selected item ID" and a list of all item IDs from the store
- Parent renders all list item children and passes in `selected` and `itemId` as props
- Child list item components are wrapped in `React.memo()` to avoid re-renders if props haven't changed
- Child list items read their own data from the store by ID

That way:

- List parent re-renders because it got the new `selectedItemId`
- List item 4 re-renders because it sees `selected` change, as does list item 7
- All other components get the same props and don't re-render

That's the hypothetical ideal. I don't know this area of the code yet, so I don't know whether it's _necessary_ here, or if we're even trying to optimize rendering perf around this part of the UI atm.

Reply:
https://www.loom.com/share/a1e09049f49742f39879e22322d18f05